### PR TITLE
generator: vision nims

### LINKED ISSUE #691: generator: multimodal nim
add new text+image -> text model

intro: https://azure.microsoft.com/en-us/blog/introducing-phi-3-redefining-whats-possible-with-slms/

api endpoint: https://build.nvidia.com/microsoft/phi-3-vision-128k-instruct

garak integration guide: https://reference.garak.ai/en/latest/contributing.generator.html

### PR DESCRIPTION
**Pull request by leondz**
resolves #691 

* New class `nim.NIMVision`, following the `huggingface.LLaVa` pattern
* New interstitial method in NIM.NVOpenAIChat that converts prompts to LLM input
* Relax a typing constraint in `attempt` to allow non-string prompts
* Specific messages for one exception type in `OpenAICompatible`

## Verification

- [ ] `garak -m nim.NIMVision -n microsoft/phi-3-vision-128k-instruct -p visual_jailbreak.FigStep -g 1 --parallel_attempts 16`
- [ ] **Verify** the thing does not do what it should not
- [ ] **Document** check the class doc in nim.NIMVision

Questions:

* Are we OK with the class name?
* Are we OK with the treatment of the oversized files in `NIMVision._prepare_prompt()` ?
* We need a pattern for modality matching/upgrading -- `huggingface.LLaVa` now needs strict matching. Tagging #602 for good and #658.

### CODE REVIEW

File: garak/attempt.py [hunk 1]
```diff
@@ -174,7 +174,7 @@ def __setattr__(self, name: str, value: Any) -> None:
         if name == "prompt":
             if value is None:
                 raise TypeError("'None' prompts are not valid")
-            assert isinstance(value, str)
+            # assert isinstance(value, str)
```
Comment (erickgalinkin) [inline garak/attempt.py:177]: Do we want to remove this line entirely instead of commenting out?

**Review by erickgalinkin — APPROVED** [review]
LGTM

File: garak/attempt.py (hunk 1 above)
Comment (jmartin-tech) [reply garak/attempt.py:177]: +1 on removal

File: garak/harnesses/base.py [hunk 2]
```diff
@@ -29,7 +29,9 @@ class Harness(Configurable):
 
     active = True
 
-    DEFAULT_PARAMS = {}
+    DEFAULT_PARAMS = {
+        "strict_modality_match": False,
+    }
```
Comment (jmartin-tech) [inline garak/harnesses/base.py:34]: This seems reasonable to be in the harness for now, however I see some possible contention as noted in the description questions.

I suspect strict modality matching requirements may lean toward to be a responsibility of the probe to define requirements. I suspect the technique employed in the probe will set the requirement for strict match as generators may often support various modalities.

**Review by jmartin-tech — APPROVED** [review]
Test look good.

``` shell
arak LLM vulnerability scanner v0.9.0.16.post1 ( https://github.com/leondz/garak ) at 2024-10-28T16:17:57.248817
📜 logging to /home/jemartin/.local/share/garak/garak.log
🦜 loading generator: NIM: microsoft/phi-3-vision-128k-instruct
📜 reporting to /home/jemartin/.local/share/garak/garak_runs/garak.ba3fbd2c-70fd-459f-b938-890b14609e6c.report.jsonl
🕵️  queue of probes: visual_jailbreak.FigStep
visual_jailbreak.FigStep                                                    visual_jailbreak.FigStep: FAIL  ok on   65/  70   (failure rate:   7.14%)
📜 report closed :) /home/jemartin/.local/share/garak/garak_runs/garak.ba3fbd2c-70fd-459f-b938-890b14609e6c.report.jsonl
📜 report html summary being written to /home/jemartin/.local/share/garak/garak_runs/garak.ba3fbd2c-70fd-459f-b938-890b14609e6c.report.html
✔️  garak run complete in 135.21s
```

``` shell
garak LLM vulnerability scanner v0.9.0.16.post1 ( https://github.com/leondz/garak ) at 2024-10-28T16:22:59.141907
📜 logging to /home/jemartin/.local/share/garak/garak.log
🦜 loading generator: NIM: microsoft/phi-3-vision-128k-instruct
📜 reporting to /home/jemartin/.local/share/garak/garak_runs/garak.df5ba817-adcb-4660-9c6b-e5507e601dab.report.jsonl
🕵️  queue of probes: encoding.InjectBase64
encoding.InjectBase64                                                           encoding.DecodeMatch: PASS  ok on   55/  55
📜 report closed :) /home/jemartin/.local/share/garak/garak_runs/garak.df5ba817-adcb-4660-9c6b-e5507e601dab.report.jsonl
📜 report html summary being written to /home/jemartin/.local/share/garak/garak_runs/garak.df5ba817-adcb-4660-9c6b-e5507e601dab.report.html
✔️  garak run complete in 7.94s
```

File: garak/harnesses/base.py (hunk 2 above)
Comment (leondz) [reply garak/harnesses/base.py:34]: Interested to hear more

**Pull request merged**